habitat-lab/habitat/tasks/rearrange/actions/actions.py

```diff
@@ -469,6 +469,163 @@ def step(self, *args, is_last_action, **kwargs):
+        front_cur_pos = trans.transform_point(
+            np.array([self._cylinder_deviate, 0, 0])
+        )
+        front_goal_pos = target_trans.transform_point(
+            np.array([self._cylinder_deviate, 0, 0])
+        )
+        rear_cur_pos = trans.transform_point(
+            np.array([-self._cylinder_deviate, 0, 0])
+        )
+        rear_goal_pos = target_trans.transform_point(
+            np.array([-self._cylinder_deviate, 0, 0])
+        )
+
+        # For step filter of the front pos
+        end_front_pos = self._sim.step_filter(front_cur_pos, front_goal_pos)
+        end_front_pos[1] = front_cur_pos[1]
+        front_goal_pos[1] = front_cur_pos[1]
+        # For step filter of the rear pos
+        end_rear_pos = self._sim.step_filter(rear_cur_pos, rear_goal_pos)
+        end_rear_pos[1] = front_cur_pos[1]
+        rear_goal_pos[1] = front_cur_pos[1]
+        # For step filter of the center pos
+        end_pos[1] = trans.translation[1]
+
+        # Robot moving distance
+        front_move = (end_front_pos - front_goal_pos).length()
+        rear_move = (end_rear_pos - rear_goal_pos).length()
+
+        # For detection of linear or angualr velocity
+        center_move = (end_pos - trans.translation).length()
+
+        # If the acutal target and the resulting moving position
+        # is too samll, then there is a collision
+        if (
+            front_move > self._lin_collision_threshold
+            or rear_move > self._lin_collision_threshold
+        ) and center_move != 0:
+            return True
+        elif (
+            front_move > self._ang_collision_threshold
+            or rear_move > self._ang_collision_threshold
+        ) and center_move == 0:
+            return True
+        else:
+            return False
+
+    def update_base(self):
+        ctrl_freq = self._sim.ctrl_freq
+
+        # before_trans_state = self._capture_articulated_agent_state()
+
+        trans = self.cur_articulated_agent.sim_obj.transformation
+        rigid_state = habitat_sim.RigidState(
+            mn.Quaternion.from_matrix(trans.rotation()), trans.translation
+        )
+
+        target_rigid_state = self.base_vel_ctrl.integrate_transform(
+            1 / ctrl_freq, rigid_state
+        )
+        end_pos = self._sim.step_filter(
+            rigid_state.translation, target_rigid_state.translation
+        )
+
+        end_pos -= self.cur_articulated_agent.params.base_offset
+        target_trans = mn.Matrix4.from_(
+            target_rigid_state.rotation.to_matrix(), end_pos
+        )
+        self.cur_articulated_agent.sim_obj.transformation = target_trans
+
+        did_coll = self.spot_collison_check(trans, target_trans, end_pos)
+
+        if not self._allow_dyn_slide:
+            # Check if in the new articulated_agent state the arm collides with anything.
+            # If so we have to revert back to the previous transform
+            self._sim.internal_step(-1)
```

Comment:
Shouldn't be stepping the simulator here.
This is general comment across all robot code as I assume this is being done elsewhere also.
If we need to reset the collision cache or something we should instead use a discrete collision detection. If there is another reason to do this, it should be factored out so we don't advance sim time arbitrarily.